### PR TITLE
Openephys

### PR DESCRIPTION
**Pull request by caniko**
This merge aims to make this repository more general purpose. This means that several functions have been moved to other repositories, or removed completely. The reviewers are welcome to put functions they believe belong to the repository back.

Tests have been added to new funcitons.

### CODE REVIEW

File: visualstimulation/plot.py [hunk 1]
```diff
@@ -1,96 +1,44 @@
...
 import matplotlib.pyplot as plt
+import matplotlib
+import numpy as np
+import seaborn as sns
 
+from .utils import make_orientation_trials
 
-def polar_tuning_curve(orients, rates, ax=None, params={}):
+
+def polar_tuning_curve(orients, rates, ax=None, transperancy=0.5, params={}):
```
Comment (alejoe91) [inline visualstimulation/plot.py]: transparency

File: visualstimulation/plot.py [hunk 2]
```diff
@@ -1,96 +1,44 @@
...
-    if spontan_rate is not None:
-        ax1.plot(orients, rates - spontan_rate, "-o", label="without bkg")
-        ax1.legend()
-
-    fig.tight_layout()
-
-    return fig
-
-
-def orient_raster_plots(trials):
```
Comment (alejoe91) [inline visualstimulation/plot.py:67]: why is this removed? It's quite convenient to have

Comment (caniko) [reply visualstimulation/plot.py:67]: It was discussed in the meeting that specific applications of the plotting functions shall be moved to the repository belonging to its respective project. I have moved the function to another [repository](https://github.com/caniko2/visual_analysis), link to [function](https://github.com/caniko2/visual_analysis/blob/889cc9e2606742f7306efcc2fce5fe0bc11af113/vian/analysis.py#L14).

It could be used as a test. Is that what you are thinking?

Comment (malinroe) [reply visualstimulation/plot.py:67]: 👍

File: examples/open_ephys_psychopy_example.ipynb [hunk 3]
```diff
@@ -0,0 +1,233 @@
+{
```
Comment (miladh) [inline examples/open_ephys_psychopy_example.ipynb]: It is really good to have notebooks with examples, but this notebook is not well documented and has some errors at the end. If it was used for testing, please remove it, otherwise add some more documentation.

File: visualstimulation/__init__.py [hunk 4]
```diff
@@ -0,0 +1,2 @@
+from .utils import *
```
Comment (miladh) [inline visualstimulation/__init__.py]: I believe it is best to avoid import "all". The chance of future changes breaking everything is higher with import*

File: visualstimulation/analysis.py [hunk 5]
```diff
@@ -36,7 +37,7 @@ def compute_dsi(rates, orients):
     out : float
         direction selectivity index
     """
-    from visualstimulation.helper import wrap_angle, find_nearest
+    from .helper import wrap_angle, find_nearest
```
Comment (miladh) [inline visualstimulation/analysis.py]: Use absolute imports instead, e.g. visualstimulation.utils.  Absolute imports are more portable and usually more readable. Also relative imports are highly discouraged by PEP8.

File: visualstimulation/analysis.py [hunk 6]
```diff
@@ -62,29 +63,34 @@ def compute_osi(rates, orients):
...
         array of orientations
+    normalise : bool
+        Feature scaling; true or false
     Returns
     -------
     out : float
         orientation selectivity index
     """
-    from visualstimulation.helper import wrap_angle, find_nearest
+    from .helper import wrap_angle, find_nearest
```
Comment (miladh) [inline visualstimulation/analysis.py]: use absolute import

File: visualstimulation/analysis.py [hunk 7]
```diff
@@ -62,29 +63,34 @@ def compute_osi(rates, orients):
...
+
+    if normalise is True:
+        n_rates = ( (rates - rates.min()) / (rates.max() - rates.min()) ) * pq.Hz
```
Comment (miladh) [inline visualstimulation/analysis.py]: `n_rates` is not used.

File: visualstimulation/analysis.py [hunk 8]
```diff
@@ -98,9 +104,14 @@ def compute_orientation_tuning(orient_trials):
...
         sorted stimulus orientations
-    '''
+    weigh: bool; default: False
+        Use gradiently weighed data
+    weights: tuple, list; default: (1, 0.6)
+        (initial weight, last weight) ===(generate graident)==> (initial weight, ..., last weight)
+    """
     from elephant.statistics import mean_firing_rate
-    from visualstimulation.helper import convert_string_to_quantity_scalar
+    from .helper import convert_string_to_quantity_scalar
```
Comment (miladh) [inline visualstimulation/analysis.py]: abs import

File: visualstimulation/plot.py [hunk 9]
```diff
@@ -1,96 +1,44 @@
-import numpy as np
 import matplotlib.pyplot as plt
+import matplotlib
+import numpy as np
+import seaborn as sns
 
+from .utils import make_orientation_trials
```
Comment (miladh) [inline visualstimulation/plot.py]: abs import

File: visualstimulation/tests/test_analysis.py [hunk 10]
```diff
@@ -42,6 +43,24 @@ def test_compute_osi():
     rates = np.array([3, 1, 2, 1, 1, 2, 1, 1])*pq.Hz
     with pytest.warns(UserWarning):
         assert abs(compute_osi(rates, orients) - 1/5) < 1e-12
+    
+    # normalise true
+    orients = np.arange(0, 315, 45)*pq.deg
+    rates = np.array([1, 1, 2, 1, 1, 2, 1, 1])*pq.Hz
+    assert abs(compute_osi(rates, orients, normalise=True) - 1) < 1e-12
```
Comment (miladh) [inline visualstimulation/tests/test_analysis.py:50]: I don't think the tests below are affected by `normalise=True` since `n_rates` is not used

File: visualstimulation/utils.py [hunk 11]
```diff
@@ -50,7 +80,7 @@ def make_orientation_trials(trials, unit=pq.deg):
         OrderedDict with orients as keys and trials as values.
     """
     from collections import defaultdict, OrderedDict
-    from visualstimulation.helper import convert_quantity_scalar_to_string, rescale_orients, convert_string_to_quantity_scalar
+    from .helper import convert_quantity_scalar_to_string, rescale_orients, convert_string_to_quantity_scalar
```
Comment (miladh) [inline visualstimulation/utils.py]: abs import

File: visualstimulation/utils.py [hunk 12]
```diff
@@ -1,12 +1,42 @@
...
 import numpy as np
 import neo
+import os
+
+
+def generate_gradiently_weighed_data(data, weight_start=1, weight_end=0):
+    """
+    Creates weighed data using gradients from weight_start to weight_end.
+    Example:
+    In:
```
Comment (miladh) [inline visualstimulation/utils.py]: Great that you have included an example, but to keep the style consistent, please use this style: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html (`def example_generator(n)`)

File: visualstimulation/tests/test_utils.py [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+
+    A = np.array([1, 2, 3, 4, 5, 6])
+    expected_A = np.linspace(weight_start, weight_end, 6) * A
+    
+    weighed_A = generate_gradiently_weighed_data(data=A, weight_start=weight_start, weight_end=weight_end)
+
+    assert all(np.isclose(expected_A, weighed_A))
+
+
+if __name__ == "__main__":
```
Comment (miladh) [inline visualstimulation/tests/test_utils.py]: Instead of running this file to run the test(s), I strongly recommend using pytest if are not already using it (https://docs.pytest.org/en/latest/). You will then be able to run all the test or a subset of them very easily.

File: examples/open_ephys_psychopy_example.ipynb (hunk 3 above)
Comment (caniko) [reply examples/open_ephys_psychopy_example.ipynb]: We can actually remove this notebook. The package no longer has the respective functionality demonstrated in the notebook. Your notebooks should suffice

File: visualstimulation/analysis.py (hunk 7 above)
Comment (miladh) [reply visualstimulation/analysis.py]: I see in this commit e80540fbd92940f5ef32c15bcf99aeb38f257d07  that the normalization part is used several more places. Therefore this part should be moved to a separate function (e.g. in utils) that you reuse other places.

Comment (caniko) [reply visualstimulation/analysis.py]: I suggest we that switch to this [function](https://stackoverflow.com/questions/52574040/min-max-scaling-along-rows-in-numpy-array) instead

File: visualstimulation/utils.py [hunk 14]
```diff
@@ -1,12 +1,51 @@
...
+    weight_end : int, float
+        Last weight
+    Returns
+    ------
+    out : numpy.array
+    Weighed data (data * weight)
+    """
+    if not isinstance(data, type(np.empty(0))):
+        msg = "data has to be numpy.array, and not {}".format(type(data))
+        raise AttributeError(msg)
```
Comment (miladh) [inline visualstimulation/utils.py]:
```suggestion
        raise TypeError(msg)
```

File: visualstimulation/utils.py [hunk 15]
```diff
@@ -1,12 +1,51 @@
...
+
+    weights = np.linspace(weight_start, weight_end, len(data))
+    print(weights)
```
Comment (miladh) [inline visualstimulation/utils.py]:
```suggestion
```

File: visualstimulation/plot.py [hunk 16]
```diff
@@ -1,96 +1,54 @@
-import numpy as np
+from sklearn.preprocessing import minmax_scale
```
Comment (miladh) [inline visualstimulation/plot.py]: Since the normalization function is quite simple, I suggest to use the implementation you had earlier (moved to a separate function). The reason for that is because we don't have any other dependencies on sklearn and generally it is a good idea to keep the number of dependencies low.

Comment (caniko) [reply visualstimulation/plot.py]: Done

File: visualstimulation/utils.py [hunk 17]
```diff
@@ -1,12 +1,67 @@
...
+    data : numpy.array
+        0D numpy.array with data to be weighed
+    Returns
+    ------
+    out : minmax scaled ([0, 1]) numpy.array
+    """
+    if not isinstance(data, type(np.empty(0))):
+        msg = "data has to be numpy.array, and not {}".format(type(data))
+        raise TypeError(msg)
+    return (data - data.min()) / (data.max() - data.min())
```
Comment (miladh) [inline visualstimulation/utils.py]: Note that, if `data` is a quantity array then this operation will remove the unit. Consider adding the following:

```
if isinstance(data, pq.Quantity):
      return (data - data.min()) / (data.max() - data.min()) * data.units
```

**Review by miladh — APPROVED** [review]

File: visualstimulation/utils.py (hunk 17 above)
Comment (caniko) [reply visualstimulation/utils.py]: Oh, I didn't know that. Shame that it doesn't include units for algebreic operations.

Note that the normalised data is a ratio, and all ratios are dimensionless. The user of the function should note the units in the plot inside an x- or a y-lable.

I agree that we should include the suggested function in case the normalised data is used for further computation.

Comment (miladh) [reply visualstimulation/utils.py]: Well, it does not remove the unit (I was imprecise), but makes it dimensionless which is technically correct so the array will still be a quantity array.
